File: apps/light-desktop/src/api/groupManagementWire.ts
import type {
	GroupManagementOperation,
	GroupManagementOutcome,
	GroupManagementRequest,
	GroupSettingsSnapshot,
	ManagedGroupProjection,
} from "../features/groupManagement/contracts";
import type {
	GroupResolvedSpatialProjection,
	GroupSpatialSelectionMapping,
	GroupManagementErrorKind as WireGroupManagementErrorKind,
	GroupManagementOperation as WireGroupManagementOperation,
	GroupManagementRequest as WireGroupManagementRequest,
} from "./generated/light-wire";
import { decodeRecordedGroupBody } from "./groupRecordingBodyWire";
import {
	arrayAt,
	booleanAt,
	enumAt,
	exactRecordAt,
	integerAt,
	numberAt,
	recordAt,
	stringAt,
} from "./playbackWirePrimitives";
import { WireValidationError } from "./wireValidation";

export const GROUP_MANAGEMENT_ERROR_KINDS = [
	"invalid",
	"unauthorized",
	"forbidden",
	"not_found",
	"conflict",
	"unavailable",
	"internal",
] as const satisfies readonly WireGroupManagementErrorKind[];

export type GroupManagementErrorKind = WireGroupManagementErrorKind;

export interface GroupManagementErrorResponse {
	kind: GroupManagementErrorKind;
	error: string;
	currentRevision: number | null;
	currentRelatedRevision: number | null;
	retryable: boolean;
}

export function encodeGroupManagementRequest(request: GroupManagementRequest) {
	printableAt(request.requestId, "$.requestId", 128, "request ID");
	printableAt(request.groupId, "$.groupId", 256, "Group ID");
	integerAt(request.expectedObjectRevision, "$.expectedObjectRevision");
	integerAt(request.expectedShowRevision, "$.expectedShowRevision");
	return {
		request_id: request.requestId,
		group_id: request.groupId,
		operation: encodeOperation(request.operation),
		expected_object_revision: request.expectedObjectRevision,
		expected_show_revision: request.expectedShowRevision,
	} satisfies WireGroupManagementRequest;
}

function encodeOperation(
	operation: GroupManagementOperation,
): WireGroupManagementOperation {
	if (operation.type === "update_properties") {
		const { name, color, icon } = operation.properties;
		printableAt(name, "$.operation.properties.name", 256, "Group name");
		return {
			type: "update_properties",
			properties: {
				name,
				color: optionalLabel(color, "$.operation.properties.color"),
				icon: optionalLabel(icon, "$.operation.properties.icon"),
			},
		};
	}
	if (operation.type === "undo") return { type: "undo" };
	if (operation.type === "set_spatial_mapping")
		return { type: "set_spatial_mapping", mapping: operation.mapping };
	if (operation.type === "remove_spatial_mapping")
		return { type: "remove_spatial_mapping" };
	const expected_source = operation.expectedSource
		? {
				source_group_id: printableAt(
					operation.expectedSource.sourceGroupId,
					"$.operation.expectedSource.sourceGroupId",
					256,
					"source Group ID",
				),
				expected_source_revision:
					operation.expectedSource.expectedSourceRevision,
			}
		: null;
	return { type: operation.type, expected_source };
}

function optionalLabel(value: string | null, path: string) {
	if (value == null) return null;
	if (new TextEncoder().encode(value).length > 64 || /\p{Cc}/u.test(value))
		invalid(path, "at most 64 printable bytes", value);
	return value;
}

export function decodeGroupManagementOutcome(
	value: unknown,
	expectedRequest: GroupManagementRequest,
): GroupManagementOutcome {
	const response = recordAt(value, "$");
	const status = enumAt(response.status, "$.status", ["changed", "no_change"]);
	const fields = [
		"status",
		"request_id",
		"correlation_id",
		"replayed",
		"show_id",
		"show_revision",
		"group",
	];
	if (status === "changed") fields.push("show_event_sequence");
	if ("persistence_warning" in response) fields.push("persistence_warning");
	exactRecordAt(response, "$", fields);
	const requestId = stringAt(response.request_id, "$.request_id");
	if (requestId !== expectedRequest.requestId)
		invalid("$.request_id", `request ${expectedRequest.requestId}`, requestId);
	const group = decodeProjection(response.group, expectedRequest);
	validateRevision(status, group.revision, expectedRequest);
	const base = {
		requestId,
		correlationId: uuidAt(response.correlation_id, "$.correlation_id"),
		replayed: booleanAt(response.replayed, "$.replayed"),
		showId: uuidAt(response.show_id, "$.show_id"),
		showRevision: integerAt(response.show_revision, "$.show_revision"),
		group,
		persistenceWarning:
			response.persistence_warning == null
				? null
				: stringAt(response.persistence_warning, "$.persistence_warning"),
	};
	if (status === "no_change") return { ...base, status };
	return {
		...base,
		status,
		eventSequence: integerAt(
			response.show_event_sequence,
			"$.show_event_sequence",
		),
	};
}

export function decodeGroupSettingsSnapshot(
	value: unknown,
	expectedGroupId: string,
): GroupSettingsSnapshot {
	const response = exactRecordAt(value, "$", [
		"show_id",
		"show_revision",
		"group",
		"resolved_spatial",
	]);
	return {
		showId: uuidAt(response.show_id, "$.show_id"),
		showRevision: integerAt(response.show_revision, "$.show_revision"),
		group: decodeProjection(response.group, expectedGroupId),
		resolvedSpatial: decodeResolvedSpatial(response.resolved_spatial),
	};
}

export function decodeGroupManagementErrorResponse(
	value: unknown,
): GroupManagementErrorResponse {
	const response = recordAt(value, "$");
	const fields = ["kind", "error", "retryable"];
	if ("current_revision" in response) fields.push("current_revision");
	if ("current_related_revision" in response)
		fields.push("current_related_revision");
	exactRecordAt(response, "$", fields);
	return {
		kind: enumAt(response.kind, "$.kind", GROUP_MANAGEMENT_ERROR_KINDS),
		error: stringAt(response.error, "$.error"),
		currentRevision: optionalInteger(
			response.current_revision,
			"$.current_revision",
		),
		currentRelatedRevision: optionalInteger(
			response.current_related_revision,
			"$.current_related_revision",
		),
		retryable: booleanAt(response.retryable, "$.retryable"),
	};
}

function optionalInteger(value: unknown, path: string) {
	return value == null ? null : integerAt(value, path);
}

function decodeProjection(
	value: unknown,
	expected: GroupManagementRequest | string,
): ManagedGroupProjection {
	const projection = recordAt(value, "$.group");
	exactRecordAt(projection, "$.group", [
		"object_id",
		"object_revision",
		"body",
	]);
	const id = printableAt(
		projection.object_id,
		"$.group.object_id",
		256,
		"Group ID",
	);
	const expectedGroupId =
		typeof expected === "string" ? expected : expected.groupId;
	if (id !== expectedGroupId)
		invalid("$.group.object_id", `Group ID ${expectedGroupId}`, id);
	const revision = integerAt(
		projection.object_revision,
		"$.group.object_revision",
	);
	return {
		id,
		revision,
		object: {
			kind: "group",
			id,
			revision,
			updated_at: "",
			body: decodeRecordedGroupBody(projection.body, id),
		},
	};
}

function decodeResolvedSpatial(value: unknown): GroupResolvedSpatialProjection {
	const spatial = exactRecordAt(value, "$.resolved_spatial", [
		"source_order",
		"effective_mapping",
		"mapping_provenance",
		"ordered_fixture_ids",
		"ranks",
		"rank_count",
		"warnings",
	]);
	return {
		source_order: uuidArray(
			spatial.source_order,
			"$.resolved_spatial.source_order",
		),
		effective_mapping:
			spatial.effective_mapping == null
				? null
				: decodeSpatialMapping(
						spatial.effective_mapping,
						"$.resolved_spatial.effective_mapping",
					),
		mapping_provenance: decodeMappingProvenance(
			spatial.mapping_provenance,
			"$.resolved_spatial.mapping_provenance",
		),
		ordered_fixture_ids: uuidArray(
			spatial.ordered_fixture_ids,
			"$.resolved_spatial.ordered_fixture_ids",
		),
		ranks: arrayAt(spatial.ranks, "$.resolved_spatial.ranks").map(
			(value, index) => {
				const rank = exactRecordAt(
					value,
					`$.resolved_spatial.ranks[${index}]`,
					["fixture_id", "rank"],
				);
				return {
					fixture_id: uuidAt(
						rank.fixture_id,
						`$.resolved_spatial.ranks[${index}].fixture_id`,
					),
					rank: integerAt(rank.rank, `$.resolved_spatial.ranks[${index}].rank`),
				};
			},
		),
		rank_count: integerAt(spatial.rank_count, "$.resolved_spatial.rank_count"),
		warnings: arrayAt(spatial.warnings, "$.resolved_spatial.warnings").map(
			(value, index) => {
				const warning = exactRecordAt(
					value,
					`$.resolved_spatial.warnings[${index}]`,
					["type", "fixture_id"],
				);
				if (warning.type !== "missing_position")
					invalid(
						`$.resolved_spatial.warnings[${index}].type`,
						"missing_position",
						warning.type,
					);
				return {
					type: "missing_position" as const,
					fixture_id: uuidAt(
						warning.fixture_id,
						`$.resolved_spatial.warnings[${index}].fixture_id`,
					),
				};
			},
		),
	};
}

function decodeMappingProvenance(value: unknown, path: string) {
	const provenance = recordAt(value, path);
	const type = enumAt(provenance.type, `${path}.type`, [
		"none",
		"local",
		"inherited",
		"mixed_source_mappings",
	]);
	if (type === "local") {
		exactRecordAt(provenance, path, ["type", "group_id"]);
		return {
			type,
			group_id: printableAt(
				provenance.group_id,
				`${path}.group_id`,
				256,
				"Group ID",
			),
		} as const;
	}
	if (type === "inherited") {
		exactRecordAt(provenance, path, ["type", "source_group_ids"]);
		return {
			type,
			source_group_ids: arrayAt(
				provenance.source_group_ids,
				`${path}.source_group_ids`,
			).map((id, index) =>
				printableAt(id, `${path}.source_group_ids[${index}]`, 256, "Group ID"),
			),
		} as const;
	}
	exactRecordAt(provenance, path, ["type"]);
	return { type } as const;
}

function decodeSpatialMapping(
	value: unknown,
	path: string,
): GroupSpatialSelectionMapping {
	const mapping = exactRecordAt(value, path, ["projection", "shape"]);
	const projection = exactRecordAt(mapping.projection, `${path}.projection`, [
		"anchor",
		"view_direction",
		"rotation_degrees",
		"preset",
	]);
	const preset =
		projection.preset == null
			? null
			: enumAt(projection.preset, `${path}.projection.preset`, [
					"top",
					"front",
					"back",
					"left",
					"right",
				]);
	return {
		projection: {
			anchor: decodePosition(projection.anchor, `${path}.projection.anchor`),
			view_direction: decodePosition(
				projection.view_direction,
				`${path}.projection.view_direction`,
			),
			rotation_degrees: numberAt(
				projection.rotation_degrees,
				`${path}.projection.rotation_degrees`,
			),
			preset,
		},
		shape: decodeShape(mapping.shape, `${path}.shape`),
	};
}

function decodePosition(value: unknown, path: string) {
	const position = exactRecordAt(value, path, ["x", "y", "z"]);
	return {
		x: numberAt(position.x, `${path}.x`),
		y: numberAt(position.y, `${path}.y`),
		z: numberAt(position.z, `${path}.z`),
	};
}

function decodeShape(
	value: unknown,
	path: string,
): GroupSpatialSelectionMapping["shape"] {
	const shape = recordAt(value, path);
	const type = enumAt(shape.type, `${path}.type`, ["grid", "radial", "radar"]);
	if (type === "grid") {
		exactRecordAt(shape, path, ["type", "angle_degrees", "direction"]);
		return {
			type,
			angle_degrees: numberAt(shape.angle_degrees, `${path}.angle_degrees`),
			direction: enumAt(shape.direction, `${path}.direction`, [
				"ascending",
				"descending",
			]),
		};
	}
	const center = {
		center_u: numberAt(shape.center_u, `${path}.center_u`),
		center_v: numberAt(shape.center_v, `${path}.center_v`),
	};
	if (type === "radial") {
		exactRecordAt(shape, path, ["type", "center_u", "center_v", "direction"]);
		return {
			type,
			...center,
			direction: enumAt(shape.direction, `${path}.direction`, [
				"outward",
				"inward",
			]),
		};
	}
	exactRecordAt(shape, path, [
		"type",
		"center_u",
		"center_v",
		"start_angle_degrees",
		"sweep",
	]);
	return {
		type,
		...center,
		start_angle_degrees: numberAt(
			shape.start_angle_degrees,
			`${path}.start_angle_degrees`,
		),
		sweep: enumAt(shape.sweep, `${path}.sweep`, [
			"clockwise",
			"counter_clockwise",
		]),
	};
}

function uuidArray(value: unknown, path: string) {
	return arrayAt(value, path).map((id, index) =>
		uuidAt(id, `${path}[${index}]`),
	);
}

function validateRevision(
	status: "changed" | "no_change",
	revision: number,
	request: GroupManagementRequest,
) {
	const expected =
		status === "changed"
			? request.expectedObjectRevision + 1
			: request.expectedObjectRevision;
	if (!Number.isSafeInteger(expected) || revision !== expected)
		invalid(
			"$.group.object_revision",
			`${status} revision ${expected}`,
			revision,
		);
}

function printableAt(
	value: unknown,
	path: string,
	byteLimit: number,
	label: string,
) {
	if (
		typeof value !== "string" ||
		!value.trim() ||
		new TextEncoder().encode(value).length > byteLimit ||
		/\p{Cc}/u.test(value)
	)
		invalid(path, `1-${byteLimit} printable ${label} bytes`, value);
	return value as string;
}

function uuidAt(value: unknown, path: string) {
	const decoded = stringAt(value, path);
	if (
		!/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(
			decoded,
		)
	)
		invalid(path, "hyphenated UUID", value);
	return decoded;
}

function invalid(path: string, expected: string, actual: unknown): never {
	throw new WireValidationError(path, expected, actual);
}

File: crates/light/adapters/headless/src/command_http/group_management_wire.rs
use light_application as application;
use light_core::Revision;
use light_wire::v2::group_management as wire;

pub(super) fn operation(
    value: wire::GroupManagementOperation,
) -> application::GroupManagementOperation {
    match value {
        wire::GroupManagementOperation::UpdateProperties { properties } => {
            application::GroupManagementOperation::UpdateProperties(
                application::GroupPropertiesUpdate {
                    name: properties.name,
                    color: properties.color,
                    icon: properties.icon,
                },
            )
        }
        wire::GroupManagementOperation::Undo {} => application::GroupManagementOperation::Undo,
        wire::GroupManagementOperation::RefreshFrozen { expected_source } => {
            application::GroupManagementOperation::RefreshFrozen {
                expected_source: expected_source.map(source_expectation),
            }
        }
        wire::GroupManagementOperation::DetachDerived { expected_source } => {
            application::GroupManagementOperation::DetachDerived {
                expected_source: expected_source.map(source_expectation),
            }
        }
        wire::GroupManagementOperation::SetSpatialMapping { mapping } => {
            application::GroupManagementOperation::SetSpatialMapping(spatial_mapping(mapping))
        }
        wire::GroupManagementOperation::RemoveSpatialMapping {} => {
            application::GroupManagementOperation::RemoveSpatialMapping
        }
    }
}

fn spatial_mapping(
    value: wire::GroupSpatialSelectionMapping,
) -> light_dynamics::SpatialSelectionMapping {
    light_dynamics::SpatialSelectionMapping {
        projection: light_dynamics::SpatialProjection {
            anchor: light_dynamics::Position3d {
                x: value.projection.anchor.x,
                y: value.projection.anchor.y,
                z: value.projection.anchor.z,
            },
            view_direction: light_dynamics::Vector3 {
                x: value.projection.view_direction.x,
                y: value.projection.view_direction.y,
                z: value.projection.view_direction.z,
            },
            rotation_degrees: value.projection.rotation_degrees,
            kind: value
                .projection
                .kind
                .map_or(light_dynamics::ProjectionKind::Planar, |kind| match kind {
                    wire::GroupMappingProjectionKind::Planar => {
                        light_dynamics::ProjectionKind::Planar
                    }
                    wire::GroupMappingProjectionKind::Cylindrical => {
                        light_dynamics::ProjectionKind::Cylindrical
                    }
                    wire::GroupMappingProjectionKind::Spherical => {
                        light_dynamics::ProjectionKind::Spherical
                    }
                }),
            axis_rotation: value.projection.axis_rotation.map_or_else(
                light_dynamics::Vector3::default,
                |rotation| light_dynamics::Vector3 {
                    x: rotation.x,
                    y: rotation.y,
                    z: rotation.z,
                },
            ),
            start_angle_degrees: value.projection.start_angle_degrees.unwrap_or(0.0),
            elevation_degrees: value.projection.elevation_degrees.unwrap_or(0.0),
            preset: value.projection.preset.map(|preset| match preset {
                wire::GroupMappingProjectionPreset::Top => light_dynamics::ProjectionPreset::Top,
                wire::GroupMappingProjectionPreset::Front => {
                    light_dynamics::ProjectionPreset::Front
                }
                wire::GroupMappingProjectionPreset::Back => light_dynamics::ProjectionPreset::Back,
                wire::GroupMappingProjectionPreset::Left => light_dynamics::ProjectionPreset::Left,
                wire::GroupMappingProjectionPreset::Right => {
                    light_dynamics::ProjectionPreset::Right
                }
            }),
        },
        shape: match value.shape {
            wire::GroupMappingShape::Grid {
                angle_degrees,
                direction,
            } => light_dynamics::SpatialSelectionShape::Grid {
                angle_degrees,
                direction: match direction {
                    wire::GroupMappingRankDirection::Ascending => {
                        light_dynamics::RankDirection::Ascending
                    }
                    wire::GroupMappingRankDirection::Descending => {
                        light_dynamics::RankDirection::Descending
                    }
                },
            },
            wire::GroupMappingShape::Radial {
                center_u,
                center_v,
                direction,
            } => light_dynamics::SpatialSelectionShape::Radial {
                center_u,
                center_v,
                direction: match direction {
                    wire::GroupMappingRadialDirection::Outward => {
                        light_dynamics::RadialDirection::Outward
                    }
                    wire::GroupMappingRadialDirection::Inward => {
                        light_dynamics::RadialDirection::Inward
                    }
                },
            },
            wire::GroupMappingShape::Radar {
                center_u,
                center_v,
                start_angle_degrees,
                sweep,
            } => light_dynamics::SpatialSelectionShape::Radar {
                center_u,
                center_v,
                start_angle_degrees,
                sweep: match sweep {
                    wire::GroupMappingRadarSweep::Clockwise => {
                        light_dynamics::RadarSweep::Clockwise
                    }
                    wire::GroupMappingRadarSweep::CounterClockwise => {
                        light_dynamics::RadarSweep::CounterClockwise
                    }
                },
            },
        },
    }
}

pub(super) fn resolved_spatial(
    snapshot: &light_engine::EngineSnapshot,
    group_id: &str,
) -> Result<wire::GroupResolvedSpatialProjection, application::ActionError> {
    let groups = snapshot
        .groups
        .iter()
        .cloned()
        .map(|group| (group.id.clone(), group))
        .collect();
    let positions = snapshot
        .dynamic_stage_positions
        .iter()
        .map(|(fixture_id, position)| {
            (
                *fixture_id,
                light_dynamics::Position3d {
                    x: f64::from(position.x),
                    y: f64::from(position.y),
                    z: f64::from(position.z),
                },
            )
        })
        .collect();
    let resolved = light_programmer::resolve_group_spatial(group_id, &groups, &positions).map_err(
        |error| application::ActionError::new(application::ActionErrorKind::Invalid, error),
    )?;
    let projected_positions = resolved.effective_mapping.as_ref().map_or_else(
        || {
            Ok(resolved
                .source_order
                .iter()
                .map(|fixture_id| wire::GroupProjectedPositionProjection {
                    fixture_id: fixture_id.0,
                    u: None,
                    v: None,
                })
                .collect())
        },
        |mapping| {
            let targets = resolved
                .source_order
                .iter()
                .map(|fixture_id| light_dynamics::SpatialTarget {
                    fixture_id: *fixture_id,
                    position: positions.get(fixture_id).copied(),
                })
                .collect::<Vec<_>>();
            light_dynamics::project_spatial_positions(&mapping.projection, &targets)
                .map(|positions| {
                    positions
                        .into_iter()
                        .map(|position| wire::GroupProjectedPositionProjection {
                            fixture_id: position.fixture_id.0,
                            u: position.u,
                            v: position.v,
                        })
                        .collect()
                })
                .map_err(|error| {
                    application::ActionError::new(
                        application::ActionErrorKind::Invalid,
                        error.to_string(),
                    )
                })
        },
    )?;
    let ranks = resolved
        .ranked_selection
        .ordered_fixture_ids
        .iter()
        .map(|fixture_id| wire::GroupSpatialRankProjection {
            fixture_id: fixture_id.0,
            rank: resolved.ranked_selection.rank_by_fixture[fixture_id],
        })
        .collect();
    Ok(wire::GroupResolvedSpatialProjection {
        source_order: resolved
            .source_order
            .iter()
            .map(|fixture_id| fixture_id.0)
            .collect(),
        effective_mapping: resolved.effective_mapping.map(wire_spatial_mapping),
        mapping_provenance: match resolved.mapping_provenance {
            light_programmer::GroupMappingProvenance::None => {
                wire::GroupMappingProvenanceProjection::None {}
            }
            light_programmer::GroupMappingProvenance::Local { group_id } => {
                wire::GroupMappingProvenanceProjection::Local { group_id }
            }
            light_programmer::GroupMappingProvenance::Inherited { source_group_ids } => {
                wire::GroupMappingProvenanceProjection::Inherited { source_group_ids }
            }
            light_programmer::GroupMappingProvenance::MixedSourceMappings => {
                wire::GroupMappingProvenanceProjection::MixedSourceMappings {}
            }
        },
        ordered_fixture_ids: resolved
            .ranked_selection
            .ordered_fixture_ids
            .iter()
            .map(|fixture_id| fixture_id.0)
            .collect(),
        projected_positions,
        ranks,
        rank_count: resolved.ranked_selection.rank_count,
        warnings: resolved
            .ranked_selection
            .warnings
            .into_iter()
            .map(|warning| match warning {
                light_dynamics::SpatialMappingWarning::MissingPosition { fixture_id } => {
                    wire::GroupSpatialWarningProjection::MissingPosition {
                        fixture_id: fixture_id.0,
                    }
                }
            })
            .collect(),
    })
}

fn wire_spatial_mapping(
    value: light_dynamics::SpatialSelectionMapping,
) -> wire::GroupSpatialSelectionMapping {
    let is_planar = value.projection.kind == light_dynamics::ProjectionKind::Planar;
    wire::GroupSpatialSelectionMapping {
        projection: wire::GroupMappingProjection {
            anchor: wire::GroupMappingPosition3d {
                x: value.projection.anchor.x,
                y: value.projection.anchor.y,
                z: value.projection.anchor.z,
            },
            view_direction: wire::GroupMappingVector3 {
                x: value.projection.view_direction.x,
                y: value.projection.view_direction.y,
                z: value.projection.view_direction.z,
            },
            rotation_degrees: value.projection.rotation_degrees,
            // A planar projection omits all four, so its payload stays exactly what it was
            // before the other kinds existed.
            kind: match value.projection.kind {
                light_dynamics::ProjectionKind::Planar => None,
                light_dynamics::ProjectionKind::Cylindrical => {
                    Some(wire::GroupMappingProjectionKind::Cylindrical)
                }
                light_dynamics::ProjectionKind::Spherical => {
                    Some(wire::GroupMappingProjectionKind::Spherical)
                }
            },
            axis_rotation: (!is_planar).then(|| wire::GroupMappingVector3 {
                x: value.projection.axis_rotation.x,
                y: value.projection.axis_rotation.y,
                z: value.projection.axis_rotation.z,
            }),
            start_angle_degrees: (!is_planar).then_some(value.projection.start_angle_degrees),
            elevation_degrees: (!is_planar).then_some(value.projection.elevation_degrees),
            preset: value.projection.preset.map(|preset| match preset {
                light_dynamics::ProjectionPreset::Top => wire::GroupMappingProjectionPreset::Top,
                light_dynamics::ProjectionPreset::Front => {
                    wire::GroupMappingProjectionPreset::Front
                }
                light_dynamics::ProjectionPreset::Back => wire::GroupMappingProjectionPreset::Back,
                light_dynamics::ProjectionPreset::Left => wire::GroupMappingProjectionPreset::Left,
                light_dynamics::ProjectionPreset::Right => {
                    wire::GroupMappingProjectionPreset::Right
                }
            }),
        },
        shape: match value.shape {
            light_dynamics::SpatialSelectionShape::Grid {
                angle_degrees,
                direction,
            } => wire::GroupMappingShape::Grid {
                angle_degrees,
                direction: match direction {
                    light_dynamics::RankDirection::Ascending => {
                        wire::GroupMappingRankDirection::Ascending
                    }
                    light_dynamics::RankDirection::Descending => {
                        wire::GroupMappingRankDirection::Descending
                    }
                },
            },
            light_dynamics::SpatialSelectionShape::Radial {
                center_u,
                center_v,
                direction,
            } => wire::GroupMappingShape::Radial {
                center_u,
                center_v,
                direction: match direction {
                    light_dynamics::RadialDirection::Outward => {
                        wire::GroupMappingRadialDirection::Outward
                    }
                    light_dynamics::RadialDirection::Inward => {
                        wire::GroupMappingRadialDirection::Inward
                    }
                },
            },
            light_dynamics::SpatialSelectionShape::Radar {
                center_u,
                center_v,
                start_angle_degrees,
                sweep,
            } => wire::GroupMappingShape::Radar {
                center_u,
                center_v,
                start_angle_degrees,
                sweep: match sweep {
                    light_dynamics::RadarSweep::Clockwise => {
                        wire::GroupMappingRadarSweep::Clockwise
                    }
                    light_dynamics::RadarSweep::CounterClockwise => {
                        wire::GroupMappingRadarSweep::CounterClockwise
                    }
                },
            },
        },
    }
}

fn source_expectation(value: wire::GroupSourceExpectation) -> application::GroupSourceExpectation {
    application::GroupSourceExpectation {
        source_group_id: value.source_group_id,
        expected_source_revision: value.expected_source_revision.map(Revision::from),
    }
}

pub(super) fn outcome(result: application::GroupManagementResult) -> wire::GroupManagementOutcome {
    let application::GroupManagementResult {
        context,
        request_id,
        replayed,
        outcome,
        persistence_warning,
    } = result;
    match outcome {
        application::GroupManagementOutcome::Changed {
            projection,
            show_revision,
            event_sequence,
        } => wire::GroupManagementOutcome::Changed {
            request_id,
            correlation_id: context.correlation_id,
            replayed,
            show_id: projection.show_id.0,
            show_revision: show_revision.value(),
            group: object_projection(&projection),
            show_event_sequence: event_sequence,
            persistence_warning,
        },
        application::GroupManagementOutcome::NoChange {
            projection,
            show_revision,
        } => wire::GroupManagementOutcome::NoChange {
            request_id,
            correlation_id: context.correlation_id,
            replayed,
            show_id: projection.show_id.0,
            show_revision: show_revision.value(),
            group: object_projection(&projection),
            persistence_warning,
        },
    }
}

fn object_projection(
    projection: &application::GroupManagementProjection,
) -> wire::GroupManagementObjectProjection {
    wire::GroupManagementObjectProjection {
        object_id: projection.object_id.clone(),
        object_revision: projection.object_revision,
        body: projection.raw_body.clone(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn transport_exposes_only_supported_management_operations() {
        assert!(matches!(
            operation(wire::GroupManagementOperation::Undo {}),
            application::GroupManagementOperation::Undo
        ));
        assert!(matches!(
            operation(wire::GroupManagementOperation::RefreshFrozen {
                expected_source: None
            }),
            application::GroupManagementOperation::RefreshFrozen {
                expected_source: None
            }
        ));
        assert!(matches!(
            operation(wire::GroupManagementOperation::DetachDerived {
                expected_source: None
            }),
            application::GroupManagementOperation::DetachDerived {
                expected_source: None
            }
        ));
    }

    #[test]
    fn a_declared_source_expectation_is_carried_exactly() {
        let application::GroupManagementOperation::RefreshFrozen {
            expected_source: Some(expectation),
        } = operation(wire::GroupManagementOperation::RefreshFrozen {
            expected_source: Some(wire::GroupSourceExpectation {
                source_group_id: "source".into(),
                expected_source_revision: Some(3),
            }),
        })
        else {
            panic!("a declared source expectation must survive translation")
        };
        assert_eq!(expectation.source_group_id, "source");
        assert_eq!(expectation.expected_source_revision, Some(3));
    }
}
